storm-client/src/jvm/org/apache/storm/utils/JCQueue.java

```diff
@@ -0,0 +1,457 @@
+        if (producerBatchSz > 1) {
+            inserter = thdLocalBatcher.get();
+            if (inserter == null) {
+                BatchInserter b = new BatchInserter(this, producerBatchSz);
+                inserter = b;
+                thdLocalBatcher.set(b);
+            }
+        } else {
+            inserter = directInserter;
+        }
+        return inserter;
+    }
+
+    /**
+     * Blocking call. Retries till it can successfully publish the obj. Can be interrupted via Thread.interrupt().
+     */
+    public void publish(Object obj) throws InterruptedException {
+        Inserter inserter = getInserter();
+        inserter.publish(obj);
+    }
+
+    /**
+     * Non-blocking call, returns false if full
+     **/
+    public boolean tryPublish(Object obj) {
+        Inserter inserter = getInserter();
+        return inserter.tryPublish(obj);
+    }
+
+    /** Non-blocking call. Bypasses any batching that may be enabled on the recvQueue. Intended for sending flush/metrics tuples */
+    public boolean tryPublishDirect(Object obj) {
+        return tryPublishInternal(obj);
+    }
+
+    /**
+     * Un-batched write to overflowQ. Should only be called by WorkerTransfer
+     * returns false if overflowLimit has reached
+     */
+    public boolean tryPublishToOverflow(Object obj) {
+        if (overflowLimit>0 && overflowQ.size() >= overflowLimit) {
+            return false;
+        }
+        overflowQ.add(obj);
+        return true;
+    }
+
+    public void recordMsgDrop() {
+        getMetrics().notifyDroppedMsg();
+    }
+
+    public boolean isEmptyOverflow() {
+        return overflowQ.isEmpty();
+    }
+
+    public int getOverflowCount() {
+        return overflowQ.size();
+    }
+
+    public int getQueuedCount() {
+        return recvQueue.size();
+    }
+
+    /**
+     * if(batchSz>1)  : Blocking call. Does not return until at least 1 element is drained or Thread.interrupt() is received
+     * if(batchSz==1) : NO-OP. Returns immediately. doesnt throw.
+     */
+    public void flush() throws InterruptedException {
+        Inserter inserter = getInserter();
+        inserter.flush();
+    }
+
+    /**
+     * if(batchSz>1)  : Non-Blocking call. Tries to flush as many as it can. Returns true if flushed at least 1.
+     * if(batchSz==1) : This is a NO-OP. Returns true immediately.
+     */
+    public boolean tryFlush()  {
+        Inserter inserter = getInserter();
+        return inserter.tryFlush();
+    }
+
+    @Override
+    public Object getState() {
+        return metrics.getState();
+    }
+
+
+    //This method enables the metrics to be accessed from outside of the JCQueue class
+    public JCQueue.QueueMetrics getMetrics() {
+        return metrics;
+    }
+
+    public interface Consumer extends org.jctools.queues.MessagePassingQueue.Consumer<Object> {
+        void accept(Object event);
+
+        void flush() throws InterruptedException;
+    }
+
+
+    public interface ExitCondition {
```

Comment:
`Predicate` with proper variable names might be able to get rid of this interface.

Reply:
yes. i had it that way originally, then felt that the call to keepRunning()  made things more readable. i can go back to predicate interface if you feel this is not desirable.

Reply:
OK makes sense. Let's keep it as it is.